docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,275 @@
+
+Enclaves provide new trust boundary protections that address old threats and open new capabilities. But just as Open Enclave and TEEs shift the security model, application developers must shift how they handle data, especially secrets, like keys and passwords.
+<br />
+<br />
+<br />
+
+## 2.1 How _Not_ to Handle Application Secrets
+Let's remind ourselves of a way _not_ to handle application secrets. A common example of this well-discussed application weakness is described by [CWE-798: Use of Hard-coded Credentials (4.4) (mitre.org)](https://cwe.mitre.org/data/definitions/798.html). While application security is often a tradeoff between multiple factors and broad edicts are not always appropriate, it's not controversial to simply say that hard-coding credentials in application code _should not be done_.
+<br />
+<br />
+<br />
+
+## 2.2 The New, Secure Enclave Way to Handle Application Secrets
+
+Enclaves provide two properties that enable applications to handle secrets securely: 1) strong identity, 2) runtime protection of secrets in memory. These two properties enable secure point-to-point secret transfers. When applications are built with enclaves, all secret management should occur dynamically _inside_ the enclave and through secure communications with trusted sources, such as cloud provider key vaults. Let's look at one example of how enclave attestation and enclave memory protection can combine to implement a secure channel.
+
+### Secure Key Release
+One of the core challenges customers have interacting with encrypted environments is how to ensure that they can securely and reliably communicate with the code running in the environment ("enclave code").
+
+One solution to this problem is what is known as "Secure Key Release", which is a pattern that enables this kind of communication with enclave code.
+
+To implement the "Secure Key Release" pattern, the enclave code generates an ephemeral asymmetric key. It then serializes the public portion of the key to some format (possibly a JSON Web Key, or PEM, or some other serialization format).
+The enclave code then calculates the SHA256 value of the public key and passes it as an input to code which generates an SGX Quote. For Open Enclave, this is done with `oe_get_evidence` or `oe_get_report`.
+
+The client then sends the SGX quote and the serialized key to the attestation service. The attestation service will validate the quote and ensure that the hash of the key is present in the quote and will issue an "Attestation Token".
+
+The client can then send that Attestation Token (which contains the serialized key) to a 3rd party "relying party". The relying party then validates that the attestation token was created by the attestation service, and thus the serialized key can be used to encrypt some data held by the "relying party" to send to the service.
```

Comment:
This statement is confusing.  Why would the Attestation Result need to contain the serialized key?   The serialized key can be obtained or negotiated via the protocol between the attester and the relying party once the RP accepts the attestation token.  Why would you want to reveal the serialized key to the attestation service?   This seems less secure (doesn't follow the need to know principle.)   Given this is a security best practices document, I think any recommendation to reveal the key to the attestation service should be removed.

Reply:
The high-level intent of this topic was to help Enclave developers envision a secure way to handle secrets and Secure Key Release was a prime model. We had not planned on covering the topic in detail here -- we thought we would simply reference existing Open Enclave SDK Secure Key Release documentation, but we could not find any. What we did find was SGX-focused documentation under the Azure Attestation docs: https://docs.microsoft.com/en-us/dotnet/api/overview/azure/security.attestation-readme#attest-sgx-enclave

The text in section 2.2 of our doc is almost verbatim from the above source, so corrections will have to be applied there also.  Would you help me rewrite this short introduction here? I'll submit a doc request for SKR against OE, and submit bugs against the multiple Azure Attestation pages that have this incorrect info.

Reply:
I'd prefer to keep the best practices doc generic (not specific to Azure) and cover "best" practices.  The azure doc you link just states it's current practice, it does not assert it is "best" practice as a model to follow, just a model.

Reply:
For the scope of this best practice guidance, I will introduce the Secure Key Release design pattern and exclude implementation details. When SKR for Open Enclave is documented, a link can be included here.

Reply:
Please see the related conversation here: https://github.com/openenclave/openenclave/pull/4178#discussion_r675261490